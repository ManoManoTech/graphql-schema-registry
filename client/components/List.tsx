import { List, ListItem, ListItemText } from '@material-ui/core';
import ChevronRightIcon from '@material-ui/icons/ChevronRight';
import React, { ReactNode, useCallback, useState } from 'react';
import {
	NavLinkProps as ReactRouterNavLinkProps,
	NavLink as ReactRouterNavLink,
} from 'react-router-dom';
import styled from 'styled-components';
import { colors } from '../utils';

export const EmptyList = styled.section`
	padding: 1rem;
	display: flex;
	justify-content: center;
	align-items: center;
	user-select: none;
`;

export const ListContainer = styled.section`
	overflow: auto;

	& > :not(:last-child) {
		border-bottom: 1px solid ${colors.black.hex256};
	}
`;

type NavigationListItemProps = {
	href: string;
	value: ReactNode;
	showNavigationChevron?: boolean;
};

export const NavigationListItem = React.forwardRef<
	HTMLAnchorElement,
	NavigationListItemProps
>(function NavigationListItem(
	{ href, value, showNavigationChevron = true },
	ref
) {
	const [active, setActive] = useState(false);
	const isActive = useCallback((match) => {
		setActive(!!match);
		return match;
	}, []);

	const CustomLink = React.useMemo(
		() =>
			React.forwardRef<
				HTMLAnchorElement,
				Omit<ReactRouterNavLinkProps, 'to'>
			>(function Link(linkProps, ref) {
				return (
					<ReactRouterNavLink
						ref={ref}
						to={href}
						isActive={isActive}
						{...linkProps}
					/>
				);
			}),
		[href, isActive]
	);

	return (
		<ListItem button component={CustomLink} selected={active} ref={ref}>
			<ListItemText primary={value} />
			{showNavigationChevron && <ChevronRightIcon />}
		</ListItem>
	);
});

type NavigationListProps = {
	children: ReactNode | ReactNode[];
};

export const NavigationList = React.forwardRef<
	HTMLElement,
	NavigationListProps
>(function NavigationList({ children }, ref) {
	return (
		<List component="nav" disablePadding ref={ref}>
			{children}
		</List>
	);
});
