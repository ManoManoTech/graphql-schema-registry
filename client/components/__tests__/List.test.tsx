import { render, screen } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import { EmptyList, NavigationList, NavigationListItem } from '../List';

describe('Lists tests', () => {
	describe('EmptyList tests', () => {
		it('should render', () => {
			render(<EmptyList />);
		});
	});

	describe('NavigationListItem tests', () => {
		it('should render', async () => {
			render(<NavigationListItem href="/test" value="Test" />, {
				wrapper: MemoryRouter,
			});

			const link = await screen.findByRole<HTMLAnchorElement>('link');

			expect(link).toBeInTheDocument();
			expect(link.href).toMatch(/.+\/test$/);
			expect(link.textContent).toBe('Test');
		});

		it('should render a chevron icon when the prop is active', async () => {
			render(
				<NavigationListItem
					href="/test"
					value="Test"
					showNavigationChevron
				/>,
				{
					wrapper: MemoryRouter,
				}
			);

			const link = await screen.findByRole<HTMLAnchorElement>('link');

			expect(link).toMatchSnapshot();
		});
	});

	describe('NavigationList tests', () => {
		it('should render', async () => {
			render(
				<NavigationList>
					<span>Dummy</span>
				</NavigationList>
			);

			const list = await screen.findByRole('navigation');

			expect(list).toBeInTheDocument();
		});
	});
});
