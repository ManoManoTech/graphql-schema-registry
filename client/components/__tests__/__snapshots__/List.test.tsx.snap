// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Lists tests NavigationListItem tests should render a chevron icon when the prop is active 1`] = `
<a
  aria-disabled="false"
  class="MuiButtonBase-root MuiListItem-root MuiListItem-gutters MuiListItem-button"
  href="/test"
  role="link"
  tabindex="0"
>
  <div
    class="MuiListItemText-root"
  >
    <span
      class="MuiTypography-root makeStyles-text-2 MuiTypography-body1 MuiTypography-displayBlock"
    >
      Test
    </span>
  </div>
  <svg
    aria-hidden="true"
    class="MuiSvgIcon-root"
    focusable="false"
    viewBox="0 0 24 24"
  >
    <path
      d="M10 6L8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z"
    />
  </svg>
  <span
    class="MuiTouchRipple-root"
  />
</a>
`;
